otc-swap-predicate/src/main.sw

```diff
@@ -0,0 +1,66 @@
+predicate;
+
+use std::{
+    address::Address,
+    assert::assert,
+    constants::BASE_ASSET_ID,
+    contract_id::ContractId,
+    option::Option,
+    result::Result,
+    revert::revert,
+    tx::{b256_from_pointer_offset, tx_input_owner, tx_output_amount, tx_output_pointer}
+};
+
+/// Order / OTC swap Predicate
+///
+/// This predicate root serves as an "order" that anyone can fill
+/// The coin sent to the predicate root can be unlocked by any transaction which has an output which meets the conditions of the order
+/// The order maker can "cancel" the order by spending the predicate's coins in any transaction containing an input they have signed
+///
+/// Limitations:
+///    - An order can not be partially filled - the taker must pay the entire ask amount
+///    - There is no on-chain matching engine, so an order placed 'offside' would not be matched with an existing order with a better price (on the contrary, it would be vulnerable to arbitrage)
+///
+/// As such, this mechanism is most useful for OTC trades and atomic swaps.
+///
+/// # Arguments
+///
+/// - `input_index` - The index of the Coin input signed by the order maker, if the order is to be cancelled.
+/// - `output_index` - The index of the Coin output which pays the order maker.
+///
+fn main(input_index: u8, output_index: u8) -> bool {
+    // Order conditions: This must be hardcoded here.
+    // The spending transaction must have an output that sends `ask_amount` of `ask_token` to `maker`
+    let maker = ~Address::from(0x0303030303030303030303030303030303030303030303030303030303030303);
+    let ask_amount = 42;
+    let ask_token = BASE_ASSET_ID;
+
+    // First, check if the transaction contains an input coin from the maker, to cancel their own order:
```

Comment:
I think it might be dangerous to only use a single owner instead of `msg_sender()`. Especially with https://github.com/FuelLabs/fuel-specs/issues/40, it would allow anyone to attach such a predicate to any partially-signed tx and potentially burn the funds if there's no change output for the given asset.

```suggestion
    // First, check if the transaction contains only input coins from the maker, to cancel their own order:
```

Reply:
I don't think we can use `msg_sender()` because of :
`^^^^^^^^ The GM (get-metadata) opcode, when called from an external context, will cause the VM to panic.`

Reply:
What do you think @adlerjohn ?

Reply:
ping @adlerjohn

Reply:
See [ff3591f](https://github.com/FuelLabs/sway-applications/pull/142/commits/ff3591f055672715c4a415c26f2e726f69da40e4)